src/pool.rs

```diff
@@ -419,74 +420,107 @@ impl ConnectionPool {
     }
 
     /// Ban an address (i.e. replica). It no longer will serve
-    /// traffic for any new transactions. Existing transactions on that replica
-    /// will finish successfully or error out to the clients.
-    pub fn ban(&self, address: &Address, client_id: i32) {
+    /// traffic for any new transactions. If this call bans the last
+    /// replica in a shard, we unban all replicas
+    pub fn ban(&self, address: &Address, client_id: i32, reason: BanReason) {
         error!("Banning {:?}", address);
         self.stats.client_ban_error(client_id, address.id);
 
-        let now = chrono::offset::Utc::now().naive_utc();
-        let mut guard = self.banlist.write();
-        guard[address.shard].insert(address.clone(), now);
-    }
+        // Primaries cannot be banned
+        if address.role == Role::Primary {
+            return;
+        }
 
-    /// Clear the replica to receive traffic again. Takes effect immediately
-    /// for all new transactions.
-    pub fn _unban(&self, address: &Address) {
-        let mut guard = self.banlist.write();
-        guard[address.shard].remove(address);
-    }
+        // We check if banning this address will result in all replica being banned
+        // If so, we unban all replicas instead
+        let pool_banned_addresses = self.ban_reporter.banlist(
+            self.settings.name.clone(),
+            self.settings.user.username.clone(),
+        );
 
```

Comment:
I moved the check for "All replicas are banned" to be in `ban` to avoid making the super hot `is_banned` method slower.